internal/sync/sync.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package sync
+
+import (
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type Mutex interface {
```

Comment:
We could use the sync.Locker and sync.RWLocker interfaces here. Should be exactly the same thing, possibly a teensy bit more documentative? I don't think there is one for the wait group though.